Add support for icons in ActionSheetIOS

matt-oakes opened the pull request:
## Summary

This PR adds support for icons next to each option in the iOS Action Sheet Picker. It allows users to specify an array of `icons` which match the indexes of the `options` which are provided. It will show the icon next to each option.

There is also a `tintIcons` option which allows switching between tinting the icon with the given color (the default behaviour) and leaving the icons as they were provided.

These props and defaults were deliberately chosen to match the [`@expo/react-native-action-sheet`](https://github.com/expo/react-native-action-sheet) library. This library uses `ActionSheetIOS` internally, so added support to React Native core would allow the props  in this library to work across Android, Web, and iOS.

## Changelog

[iOS] [Added] - Add support for icons in ActionSheetIOS

## Test Plan

I have added three new cases to RNTester and tested it manually. The screenshots below show it with icons that were already present in RNTester. It would look nicer with smaller icons, but I didn't want to find  suitable icons were had the correct licencing to include in the repo.

**Icons for first two options with default tint**

![Simulator Screen Shot - iPhone 11 Pro Max - 2020-04-16 at 09 31 40](https://user-images.githubusercontent.com/97068/79433551-2a1c3d00-7fc5-11ea-9057-dbfa60ed8a91.png)

**Icons for first two options with custom tint**

![Simulator Screen Shot - iPhone 11 Pro Max - 2020-04-16 at 09 31 44](https://user-images.githubusercontent.com/97068/79433568-30121e00-7fc5-11ea-8892-52c576204abe.png)

**Icons for first two options with tinting turned off**

![Simulator Screen Shot - iPhone 11 Pro Max - 2020-04-16 at 09 31 48](https://user-images.githubusercontent.com/97068/79433588-37d1c280-7fc5-11ea-856d-43eb6088d70d.png)

# TODO

- [x] I have manually modified the `FBReactNativeSpec.h` file to support the new options. This is actually a file which is generated by a Facebook internal version of codegen. Can someone at Facebook generate the new version of this file and commit it to this PR so it is correct?

hramos commented on Libraries/ActionSheetIOS/NativeActionSheetManager.js:
```diff
@@ -25,6 +25,8 @@ export interface Spec extends TurboModule {
       +anchor?: ?number,
       +tintColor?: ?number,
       +userInterfaceStyle?: ?string,
+      +icons: ?Array<Object>, // TODO: This should be a ResolvedAssetSource
+      +tintIcons: boolean,
```
```suggestion
      +icons?: ?Array<Object>, // TODO: This should be a ResolvedAssetSource
      +tintIcons?: boolean,
```

matt-oakes replied:
As noted in the PR description, I chose these option names to match the existing ones from the Expo action sheet library, which uses ActionSheetIOS internally. The options are named this way for Android and web and changing them would introduce some complexity and incompatabilites. Details here: https://github.com/expo/react-native-action-sheet#androidweb-only-props

hramos commented on React/CoreModules/RCTActionSheetManager.mm:
```diff
@@ -123,11 +130,21 @@ - (void)presentViewController:(UIViewController *)alertController
-                                                      handler:^(__unused UIAlertAction *action) {
-                                                        callback(@[ @(localIndex) ]);
-                                                      }]];
+    UIAlertAction *action = [UIAlertAction actionWithTitle:option style:style handler:^(__unused UIAlertAction *action) {
+      callback(@[ @(localIndex) ]);
+    }];
+      
```
```suggestion
    UIAlertAction *action = [UIAlertAction actionWithTitle:option
                                                     style:style
                                                   handler:^(__unused UIAlertAction *action) {
                                                     callback(@[ @(localIndex) ]);
                                                   }];
      
```

hramos commented on React/CoreModules/RCTActionSheetManager.mm:
```diff
@@ -239,6 +256,26 @@ - (void)presentViewController:(UIViewController *)alertController
+  if (!iconURLString) {
+    return nil;
+  }
+  
```
```suggestion

```

hramos commented on React/CoreModules/RCTActionSheetManager.mm:
```diff
@@ -239,6 +256,26 @@ - (void)presentViewController:(UIViewController *)alertController
+  if (!iconURL) {
+    return nil;
+  }
+  
```
```suggestion

```

hramos commented on React/CoreModules/RCTActionSheetManager.mm:
```diff
@@ -239,6 +256,26 @@ - (void)presentViewController:(UIViewController *)alertController
+  if (!iconData) {
+    return nil;
+  }
+  
```
```suggestion

```

hramos commented on Libraries/FBReactNativeSpec/FBReactNativeSpec/FBReactNativeSpec.h:
```diff
@@ -118,6 +118,8 @@ namespace JS {
       folly::Optional<double> tintColor() const;
       NSString *userInterfaceStyle() const;
+      folly::Optional<facebook::react::LazyVector<id<NSObject>>> icons() const;
+      bool tintIcons() const;
```
```suggestion
      folly::Optional<bool> tintIcons() const;
```

hramos commented on Libraries/FBReactNativeSpec/FBReactNativeSpec/FBReactNativeSpec.h:
```diff
@@ -2942,6 +2944,16 @@ inline NSString *JS::NativeActionSheetManager::SpecShowActionSheetWithOptionsOpt
+  id const p = _v[@"icons"];
+  return RCTBridgingToOptionalVec(p, ^id<NSObject>(id itemValue_0) { return itemValue_0; });
+}
+inline bool JS::NativeActionSheetManager::SpecShowActionSheetWithOptionsOptions::tintIcons() const
+{
+  id const p = _v[@"tintIcons"];
+  return RCTBridgingToBool(p);
+}
```
```suggestion
inline folly::Optional<bool> JS::NativeActionSheetManager::SpecShowActionSheetWithOptionsOptions::tintIcons() const
{
  id const p = _v[@"tintIcons"];
  return RCTBridgingToOptionalBool(p);
}
```

matt-oakes commented:
@hramos Thanks for the review. I've addressed all of those issues now. I've also made the `tintOptions` optional for the native module and moved the "default to `true`" logic into native code rather than JavaScript.

sota000 commented:
@matt-oakes, I know this is 1+ year old PR, but would you be wiling to rebase and solve the merge conflict? After that, I'll triage it internally. I'll check back in a few days.

matt-oakes commented:
@sota000 Thanks for looking into this. I'd actually given up hope that it was going to be looked at.

I have rebased on `main`, fixed the conflicts, and tested that everything still works in RNTester. Everything should be good to merge in now 👍

sota000 commented:
Thanks, @matt-oakes! Sorry again for the delay (missed the notification). I've imported this to review this internally. Thanks!

philIip commented on Libraries/ActionSheetIOS/NativeActionSheetManager.js:
```diff
@@ -24,6 +24,8 @@ export interface Spec extends TurboModule {
       +userInterfaceStyle?: ?string,
       +disabledButtonIndices?: Array<number>,
+      +icons?: ?Array<Object>, // TODO: This should be a ResolvedAssetSource
+      +tintIcons?: ?boolean,
```
i think it's more clear to make this a verb, i.e. `shouldTintIcons`. otherwise, i would think this is an array of icon metadata, and since we lose the type in the obj-c layer, that could be very misleading.

also, this conforms more to obj-c naming standards

philIip commented on React/CoreModules/RCTActionSheetManager.mm:
```diff
@@ -131,11 +138,23 @@ - (void)presentViewController:(UIViewController *)alertController
+          iconImage = [iconImage imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
+        }
+        [action setValue:iconImage forKey:@"image"];
+      }
```
i feel it is probably better to have an error here instead of failing silently, if the consumer is passing down a list of icons then presumably iconImage should never be nil

matt-oakes replied:
This was origionally done to allow passing in a `null` value for some of the icons, as described above. However, I think you are correct that it is better to show an error when a value has been passed in.

I have changed the code so it will ignore `null` values, but will show an error when something has been passed in.

philIip commented on Libraries/ActionSheetIOS/NativeActionSheetManager.js:
```diff
@@ -24,6 +24,8 @@ export interface Spec extends TurboModule {
       +tintColor?: ?number,
       +userInterfaceStyle?: ?string,
       +disabledButtonIndices?: Array<number>,
+      +icons?: ?Array<Object>, // TODO: This should be a ResolvedAssetSource
```
do you plan to land this code with this comment? it's not totally clear here

matt-oakes replied:
I do, because I wasn't sure how to correctly type this and there is no documentation for Turbo Modules that's publicly available yet. If there is a better way to do this, then feel free to either let me know or add a commit to this branch.

RSNara replied:
Hey @matt-oakes!

In the TurboModule codegen, you can use any Flow types that you declare in the Spec. So, if you know the shape of ResolvedAssetSource, you can do this:

```
type ResolvedAssetSource = {
  foo: string,
};

icons: ?Array<ResolvedAssetSource>
```

matt-oakes replied:
@RSNara Can you point me to a place in the code where the shape of this object is defined, or another place where it's passed through JSI?

At the time I wrote this, the other places that were passing asset references were doing what I did, however, that might have changed now.

philIip reviewed (changes requested):
thanks a lot for working on this! i have some questions / recommendations.

matt-oakes commented on React/CoreModules/RCTActionSheetManager.mm:
```diff
@@ -131,11 +138,23 @@ - (void)presentViewController:(UIViewController *)alertController
+                                                     callback(@[ @(localIndex) ]);
+                                                   }];
+      
+    if (index < icons.count) {
```
I implemented it this way so it matches the implementation of the existing Android & Web only options in the Expo Action Sheet. Details here: https://github.com/expo/react-native-action-sheet#androidweb-only-props

The intention of this pull request was to add support for these options into React Native code as the Expo library relies on ActionSheetIOS for it's iOs implementation, while providing its own implementation for Android and web.

Changing the options in the core would just complicate usage for existing users of the Expo library.

However, to support the use case you mentioned, I have added the ability to pass in `null` values into the array to signify that an icon should not be used for that index.

matt-oakes replied:
I have added a test case to RNTester which explicitly tests this use case by supplying a `null` for the first value and an icon for the 2nd

sota000 commented:
@matt-oakes, mind taking a look at the comment?

matt-oakes commented:
@sota000 I have replied to the comment.

I don't know how to fix this issue because it's something related to the internal structure of the assets which is not documented anywhere. That is why I left this as a TODO for someone with more knowledge to add in, rather than trying to guess myself.

Personally, I would say to merge this in as-is as it works and the last remaining issue is a small code quality thing that can be tidied up by someone with more knowledge in the future. The type saftey won't really add much at this point.

sota000 commented:
I'll see what we can do. Thanks for being patient!

RSNara commented:
Hey @matt-oakes, I just saw your comment. It looks like ResolvedAssetSource is defined here: https://github.com/facebook/react-native/blob/77ecc7ede1da8fc590d7bc238a2fc02daa736746/Libraries/Image/AssetSourceResolver.js#L13-L19. 

I found this with a quick search through the repository. Did you encounter this type earlier, but rule it out, because it was invalid/not applicable in this scenario for some reason?

matt-oakes commented:
@RSNara I have made that change now and rebased on `main`
